enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,1098 @@
+	// The current default value is "RollingUpdate".
+	// +kubebuilder:default:="RollingUpdate"
+	// +kubebuilder:validation:Enum:="RollingUpdate";"OnDelete"
+	// +optional
+	Type ControlPlaneMachineSetStrategyType `json:"type,omitempty"`
+
+	// This is left as a struct to allow future rolling update
+	// strategy configuration to be added later.
+}
+
+// ControlPlaneMachineSetStrategyType is an enumeration of different update strategies
+// for the Control Plane Machines.
+type ControlPlaneMachineSetStrategyType string
+
+const (
+	// RollingUpdate is the default update strategy type for a
+	// ControlPlaneMachineSet. This will cause the ControlPlaneMachineSet to
+	// first create a new Machine and wait for this to be Ready
+	// before removing the Machine chosen for replacement.
+	RollingUpdate ControlPlaneMachineSetStrategyType = "RollingUpdate"
+
+	// Recreate causes the ControlPlaneMachineSet controller to first
+	// remove a ControlPlaneMachine before creating its
+	// replacement. This allows for scenarios with limited capacity
+	// such as baremetal environments where additional capacity to
+	// perform rolling updates is not available.
+	Recreate ControlPlaneMachineSetStrategyType = "Recreate"
+
+	// OnDelete causes the ControlPlaneMachineSet to only replace a
+	// Machine once it has been marked for deletion. This strategy
+	// makes the rollout of updated specifications into a manual
+	// process. This allows users to test new configuration on
+	// a single Machine without forcing the rollout of all of their
+	// Control Plane Machines.
+	OnDelete ControlPlaneMachineSetStrategyType = "OnDelete"
+)
+
+// FailureDomain represents the different configurations required to spread Machines
+// across failure domains on different platforms.
+// +union
+type FailureDomains struct {
+	// Platform identifies the platform for which the FailureDomain represents
+	// +unionDiscriminator
+	// +optional
+	Platform configv1.PlatformType `json:"platform,omitempty"`
+
+	// AWS configures failure domain information for the AWS platform
+	// +optional
+	AWS *[]AWSFailureDomain `json:"aws,omitempty"`
+
+	// Azure configures failure domain information for the Azure platform
+	// +optional
+	Azure *[]AzureFailureDomain `json:"azure,omitempty"`
+
+	// GCP configures failure domain information for the GCP platform
+	// +optional
+	GCP *[]GCPFailureDomain `json:"gcp,omitempty"`
+
+	// OpenStack configures failure domain information for the OpenStack platform
+	// +optional
+	OpenStack *[]OpenStackFailureDomain `json:"openstack,omitempty"`
+}
+
+// AWSFailureDomain configures failure domain information for the AWS platform
+// +kubebuilder:validation:MinProperties:=1
+type AWSFailureDomain struct {
+	// Subnet is a reference to the subnet to use for this instance.
+  // If no subnet reference is provided, the Machine will be created in the first
+  // subnet returned by AWS when listing subnets within the provided availability zone.
```

Comment:
Is this the current behavior in the Machine API? It seems unlikely (1) that an account will only have a single subnet in a region and (2) that in the case of multiple subnets that the first one selected will be the one that the user wants to use.

Reply:
Yep, this is behaviour that was carried over from Cluster API when we forked from that initially, not sure this behaviour is great, but we can't change it at this point

Reply:
Are we stuck on perpetuating that to the ControlPlaneMachineSet? Do we not have the choice to make this subnets field mandatory? Or does that go back to diverging too much from Cluster API, with respect to the possibility of merging back with them at some point in the future? Could we start as required and make the field optional in the future?

Reply:
> Are we stuck on perpetuating that to the ControlPlaneMachineSet?

I had intended to keep the behaviour the same so that we don't have to have caveats in our documentation such as "this is required for Control Plane Machines but optional for Worker Machines"

> Could we start as required and make the field optional in the future?

I've asked API folks about this in the past, changing required to optional can cause issues for API consumers. If we define an API field as required, then the consumer doesn't need to handle when it's not set, changing it to optional later would then break their logic, so this in general is not allowed

> Or does that go back to diverging too much from Cluster API, with respect to the possibility of merging back with them at some point in the future?

Cluster API have optional subnets on their Machines so ideally we stick to that too. In their case however, there are a list of subnets on the `AWSCluster` resource which they fall back to when it's not provided, so it's a little better in terms of the behaviour.

IMO at the moment at least, we should leave this optional and retain the same behaviour between control plane and worker machines. All of our docs examples include the subnets already and I would expect users to start their CPMS config based on their existing control plane machines, which will have the subnets defined. 
If they don't define the subnets, then the Machines may end up in the correct subnet, or they won't, and if they don't, they will likely never become running because the traffic between the new node and the rest of the cluster won't be allowed. In this case, our new CPMS operator will go degraded and highlight to them that there is an issue

Reply:
Is the subnet set in the Machine somewhere? Are we covered in a scenario where there are subnets added to the account *after* the ControlPlaneMachineSet is created with omitted subnets? For example, if a machine needs to be re-created, can we guarantee that the new machine is added to the same subnet as the existing machines?

Reply:
>  > Is the subnet set in the Machine somewhere?

> Yes

Sorry, I was not clear enough here. What I meant to ask was whether the subnet chosen by the machine-api when the user omits the subnet would be stored in the Machine resource.

> I would expect when we integrate this with the installer, it will set the subnets by filter anyway, so this is only in theory an issue for users who install this as a day 2

Right. I don't expect this to be an issue for the vast, vast majority of users. However, that again begs the question of why we would want to even offer the option to configure it in a way that has the high likelihood of failing. I get that we are trying to be consistent between APIs. There is value in that. But we should also consider whether we want to perpetuate deficiencies in an API solely for the sake of consistency. If that consideration has been made--which it seems to have been--and the conclusion is that we do prefer consistency in this case, then I can live with that. I do not share the same opinion, but that is fine.

Reply:
> Sorry, I was not clear enough here. What I meant to ask was whether the subnet chosen by the machine-api when the user omits the subnet would be stored in the Machine resource.

Oh got it, no we don't store it anywhere

> If that consideration has been made--which it seems to have been--and the conclusion is that we do prefer consistency in this case, then I can live with that. I do not share the same opinion, but that is fine.

It has previously been discussed within the team, but I can bring it up again in case opinions have changed. In the prior conversation I think the big sticking point was that we don't want to block a user from doing something on the control plane that they in theory can do today with an unmanaged control plane machine. If we were to do that, then there's a probably very small chance that someone might have to circumvent the management we are putting in, at which point this proposal adds negative value for them.

I think the previous conversation was around options and fields in general (and trying to create a blanket rule) rather than specifically the subnets, so I'll bring this up with the team and see if anyone has a changed opinion if we look directly at the subnet case

Reply:
I discussed this with the team and for now, we prefer to keep the consistency with the worker and control plane machines even in the case of the subnets, going to resolve this